x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go

```diff
@@ -0,0 +1,194 @@
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	clienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	connectiontypes "github.com/cosmos/ibc-go/v3/modules/core/03-connection/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	commitmenttypes "github.com/cosmos/ibc-go/v3/modules/core/23-commitment/types"
+	ibctmtypes "github.com/cosmos/ibc-go/v3/modules/light-clients/07-tendermint/types"
+
+	// host "github.com/cosmos/ibc-go/v3/modules/core/24-host"
+	ibctesting "github.com/cosmos/ibc-go/v3/testing"
+
+	_ "github.com/stretchr/testify/suite"
+)
+
+type ClaimUndelegatedTestCase struct {
+	redemptionRecordId string
+	validMsg           stakeibc.MsgClaimUndelegatedTokens
+	expectedTxMsg      stakeibckeeper.IcaTx
+	redemptionRecord   recordtypes.UserRedemptionRecord
+	hostZone           types.HostZone
+}
+
+func (s *KeeperTestSuite) SetupClaimUndelegatedTokens() ClaimUndelegatedTestCase {
+	senderAddr := "stride_SENDER"
+	receiverAddr := "cosmos_RECEIVER"
+	redemptionAddr := "cosmos_REDEMPTION"
+	chainId := "GAIA"
+	redemptionRecordId := "GAIA.1.stride_SENDER"
+	connectionId := "connection-0"
+
+	redemptionAccount := stakeibc.ICAAccount{
+		Address: redemptionAddr,
+		Target:  stakeibc.ICAAccountType_REDEMPTION,
+	}
+	hostZone := stakeibc.HostZone{
+		ChainId:           chainId,
+		RedemptionAccount: &redemptionAccount,
+		ConnectionId:      connectionId,
+	}
+
+	redemptionRecord := recordtypes.UserRedemptionRecord{
+		Id:          redemptionRecordId,
+		HostZoneId:  chainId,
+		EpochNumber: 1,
+		Sender:      senderAddr,
+		Receiver:    receiverAddr,
+		Denom:       "uatom",
+		IsClaimable: true,
+		Amount:      1000,
+	}
+	redemptionAmount := sdk.NewCoins(sdk.NewInt64Coin(redemptionRecord.Denom, int64(redemptionRecord.Amount)))
+
+	epochTracker := stakeibc.EpochTracker{
+		EpochIdentifier:    epochtypes.STRIDE_EPOCH,
+		EpochNumber:        1,
+		NextEpochStartTime: 0,
+	}
+
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, hostZone)
+	s.App.StakeibcKeeper.SetEpochTracker(s.Ctx, epochTracker)
+	s.App.RecordsKeeper.SetUserRedemptionRecord(s.Ctx, redemptionRecord)
+
+	connectionEnd := connectiontypes.ConnectionEnd{ClientId: "client-0", Versions: []*connectiontypes.Version{connectiontypes.DefaultIBCVersion}}
+	owner := stakeibc.FormatICAAccountOwner("STRIDE", stakeibc.ICAAccountType_REDEMPTION)
+	portId := "icacontroller-" + owner
+	state := ibctmtypes.NewClientState("STRIDE", ibctmtypes.DefaultTrustLevel, ibctesting.TrustingPeriod, ibctesting.UnbondingPeriod, ibctesting.MaxClockDrift, clienttypes.NewHeight(0, 10), commitmenttypes.GetSDKSpecs(), ibctesting.UpgradePath, false, false)
+	s.App.StakeibcKeeper.IBCKeeper.ClientKeeper.SetClientState(s.Ctx, "client-0", state)
+	s.App.StakeibcKeeper.IBCKeeper.ConnectionKeeper.SetConnection(s.Ctx, connectionId, connectionEnd)
+	s.App.StakeibcKeeper.IBCKeeper.ConnectionKeeper.SetConnection(s.Ctx, connectionId, connectionEnd)
```

Comment:
nice job figuring this stuff out, it's not easy

Reply:
is this line duplicated?

Reply:
sorry all that stuff was very much in "debugging state". I just wanted to wait to remove it until the tests were done to be safe. Doing it now - will push a commit in a sec!